doc/develop/languages/cpp/index.rst

```diff
@@ -121,5 +121,133 @@ library can select, from its config,
+files: data structures, macros, static functions,...  While C and C++
+have a large overlap, they're different languages with `known
+incompatibilities`_. C is not just a C++ subset. Standard levels (e.g.:
+"C+11") add another level of complexity as new features are often
+inspired by and copied from the other language but many years later and
+with subtle differences. Making things more complex, compilers often
+offer early prototypes of features before they become
+standardized. Standards can have ambiguities interpreted differently by
+different compilers. Compilers can have bugs and these may need
+workarounds. To help with this, many projects restrict themselves to a
+limited number of toolchains. Zephyr does not.
+
+These compatibility issues affect header files dis-proportionally.  Not
+just because they have to be compatible between C and C++, but also
+because they end up being compiled in a surprisingly high number of other
+source files due to *indirect* inclusion and the `lack of structure and
+headers organization`_ that is typical in real-world projects. So, header
+files are exposed to a much larger variety of toolchains and project
+configurations.
+Adding more constraints, the Zephyr project has demanding policies
+with respect to code style, compiler warnings, static analyzers and
+standard compliance (e.g.: MISRA).
+
+Put together, all these constraints can make writing header files very
+challenging. The purpose of this section is to document some best "header
+practices" and lessons learned in a Zephyr-specific context. While a lot
+of the information here is not Zephyr-specific, this section is not a
+substitute for knowledge of C/C++ standards, textbooks and other
+references.
+
+Testing
+-------
+
+Fortunately, the Zephyr project has an extensive test and CI
+infrastructure that provides coverage baselines, catches issues early,
+enforces policies and maintains such combinatorial explosions under some
+control. The ``tests/lib/cpp/cxx/`` are very useful in this context
+because their ``testcase.yaml`` configuration lets ``twister`` iterate
+quickly over a range of ``-std`` parameters: ``-std=c++98``,
+``-std=c++11``, etc.
+
+Keep in mind unused macros are not compiled.
+
+Designated initializers
+-----------------------
+
+Initialization macros are common in header files as they help reduce
+boilerplate code. C99 added initialization of ``struct`` and ``union``
+types by "designated" member names instead of a list of *bare*
+expressions. Some GCC versions support designated initializers even in
+their C90 mode.
+
+When used at a simple level, designated initializers are less
+error-prone, more readable and more flexible. On the other hand, C99
+allowed a surprisingly large and lax set of possibilities: designated
+initializers can be out of order, duplicated, "nested" (``.a.x =``),
+various other braces can be omitted, designated initializers and not can
+be mixed, etc.
```

Comment:
If there's concern for concision, this paragraph could be dropped.  Users coming to Zephyr with these initializers already know how they work and why they're good, and if they don't they won't care.  We don't need to document C, just our own environment.

Reply:
I considered this as a relatively short introduction to the main topic... I agree we don't need to document C but I think having a few examples of incompatibilities upfront helps.